### ISSUE TITLE
OSIO Launcher - New Error State

### ISSUE THREAD

**Issue by bdellasc** [openshiftio/openshift.io#3152]
There's an error that can came up during the click party where the user clicks the "Setup Application" button at the end of the launcher flow, but the system has an issue before it can ever get to the status page. 

I asked them to send some screen caps of it. They wanted to know how to handle that situation. My gut was telling me to add the error to the Summary section, so they don't leave that UI and can just retry the setup again from the Summary page.

**Comment by invincibleJai** [comment]
@bdellasc one sample scenario is : 

- When user tries to `setup app`, some input data may be wrong i.e 400 bad request/ 500 internal server error. ( these are not so common scenario but we need to had to handle it in generic way)

<img width="1670" alt="screen shot 2018-04-18 at 4 53 07 pm" src="https://user-images.githubusercontent.com/5129024/38929250-58848b00-4329-11e8-886a-a6f1dca973a9.png">

<img width="1671" alt="screen shot 2018-04-18 at 4 53 27 pm" src="https://user-images.githubusercontent.com/5129024/38929258-62b2efc2-4329-11e8-9129-dfeb61f183e6.png">

**Comment by bdellasc** [comment]
@invincibleJai - Before I mock anything up, here's what I would expect to happen:

1. User clicks the blue "Set Up Application" button
2. Error 400 or 500 triggers, and we show a custom error message with a blue primary action button to return the user to the Launcher summary section (they could also click the "Back" button in the browser).
3. Because the information from the Laucher UI was cached in the browser, they'll be brought back to the summary screen with all their selections populated and the blue "Set Up Application" button.

If my assumption is wrong about the user's Launcher UI information still being preserved in the browser cache, please let me know. If what I described above sounds accurate from a technical standpoint, I'll put together a couple of mockups.

tagging @mindreeper2420 , as we were discussing this earlier.

**Comment by jyasveer** [comment]
@bdellasc i doubt we can go back to the same state as we were in Summary screen. I have to see the feasibility here

**Comment by invincibleJai** [comment]
@bdellascI don't think it's an ask from GA point of view. I highlighted it just to have a better user experience as in case of some invalid input/error API call will silently fail at least in those scenarios user to be prompted with a notification saying "there is an issue with repo name entered by you or groupID etc".

> 1. User clicks the blue "Set Up Application" button
> 2. Error 400 or 500 triggers, then we show a custom error message with a blue primary action button to return the user to the Launcher summary section. The user could also click the "Back" button in the browser.

why not keep user in summary section itself and show toast notification with error updating user there is issue with this data , pls correct.

> 3. Because the information from the Laucher UI was cached in the browser, they'll return to the summary screen with all their selections populated and the blue "Set Up Application" button available to click again.

there won't be any need for user to come back as user would be in summary screen only, he can choose to correct invalid info and click "Set Up Application" button

CC @jyasveer

**Comment by bdellasc** [comment]
I spoke with @dlabrecq and he had some insights on the challenges around this issue. I'll let him chime in. Please keep the conversation/progress updates in this comment thread so it's all in one place. Thanks!

**Comment by bdellasc** [comment]
@invincibleJai @jyasveer @dlabrecq I think Dan was suggesting a similar approach to what you described above. My initial proposal was a shot in the dark without a lot of background knowledge on what was technically feasible. If we're all ok with the toast notification approach, I'm fine with that as well. 

I think we'd just need to make a pass at the content for that toast notification, and for that, I'd need to just add some info to this story to reflect that work. Dan, let us know if @invincibleJai 's proposal aligns with what you had in mind.

**Comment by dlabrecq** [comment]
It's not clear when the 400/500 error is received. Is it before landing on the progress page?

That said, my preference would be to keep the user in the summary screen (as @invincibleJai suggested) and display a toast notification or inline error.

Looking at the screenshot, I see messages like "pipeline is required", "runtime is required", etc.  I assume this is due to missing/incorrect summary data? We should be able to verify the data is correct before proceeding? The "Set Up Application" button should be disabled until then.

FYI, it is possible to restore state if necessary. The browser does not cache the data, but we already restore state when the user navigates away from the app-launcher to authenticate with GitHub. As long as you provide the necessary query params in the app-launcher's URL, all state can be restored.

**Comment by invincibleJai** [comment]
> It's not clear when the 400/500 error is received. Is it before landing on the progress page?

yes Dan it's before landing on the progress page, for progress screen we do have error notification.

> Looking at the screenshot, I see messages like "pipeline is required", "runtime is required", etc. I assume this is due to missing/incorrect summary data? We should be able to verify the data is correct before proceeding? The "Set Up Application" button should be disabled until then.

@dlabrecq agree with you that's the expectation, I came across this scenario a few days back when GitHub repo name didn't have validation apart from check if repo name exists/ doesn't exist so during click part one user has entered repo with space in it i.e "my repo" and it silently failed in summary screen with 400. Though we have the validation in place now, this is one such example.

The idea here is to have some notification for the user if it fails due to any reason.

**Comment by bdellasc** [comment]
Given the discussion, is there enough information to implement what @dlabrecq suggested in his last post? I don't get a sense that you'd need anything from a mockup standpoint to do what he described?

**Comment by bdellasc** [comment]
@invincibleJai @mindreeper2420 @jyasveer Where are we with this issue? Is anything needed from me?

**Comment by invincibleJai** [comment]
@bdellasc with the approach suggested we should good. We didn't get time to visit this as were engaged in other tasks.

CC @animuk

**Comment by bdellasc** [comment]
I will look into the original dev story and update the comments in this story accordingly.
